big_vision/trainers/proj/gsam/train.py

```diff
@@ -0,0 +1,401 @@
+  # be sent there later as needed, otherwise we already encountered two
+  # situations where we allocate them twice.
+  @partial(jax.jit, backend="cpu")
+  def init(rng):
+    shape = tuple(train_ds.element_spec["image"].shape[1:])
+    bs = config.batch_size // jax.device_count()
+    dummy_input = jnp.zeros((bs,) + shape, jnp.float32)
+    params = flax.core.unfreeze(model.init(rng, dummy_input))["params"]
+
+    # Set bias in the head to a low value, such that loss is small initially.
+    if "init_head_bias" in config:
+      params["head"]["bias"] = jnp.full_like(params["head"]["bias"],
+                                             config["init_head_bias"])
+
+    return params
+
+  rng, rng_init = jax.random.split(rng)
+  params_cpu = init(rng_init)
+
+  if jax.process_index() == 0:
+    num_params = sum(p.size for p in jax.tree_leaves(params_cpu))
+    parameter_overview.log_parameter_overview(params_cpu, msg="init params")
+    mw.measure("num_params", num_params)
+
+  write_note(f"Initializing {config.optax_name} optimizer...")
+  tx, sched_fns = bv_optax.make(config, params_cpu, sched_kw=dict(
+      global_batch_size=batch_size,
+      total_steps=total_steps,
+      steps_per_epoch=steps_per_epoch))
+
+  # We jit this, such that the arrays are created on the CPU, not device[0].
+  opt_cpu = jax.jit(tx.init, backend="cpu")(params_cpu)
+  sched_fns_cpu = [jax.jit(sched_fn, backend="cpu") for sched_fn in sched_fns]
+
+  # ----------------------------------------------------------------------------
+  # Get hyper-params for GSAM.
+  RHO_MAX, RHO_MIN = config.get("rho_max", 0.3), config.get("rho_min", 0.05)
+
+  warnings.warn("Please make sure 'config.lr_min' equals the min value by your lr_schedule. \
+    Perturbation amplitude schedule is: \
+    \n sam_rho = rho_min + (rho_max - rho_min) * (lr - lr_min) / (lr_max - lr_min) ")
+
+  LR_MAX = config.get("lr", 3e-3)
+  LR_MIN = config.get("lr_min", LR_MAX * 0.01)
+
+  ALPHA = config.get("alpha", 0.05)
+  ADAPTIVE_PERTURBATION = config.get("adaptive_perturbation", False)
+  MINIMIZE_FP = config.get("minimize_fp", True)
+  # ----------------------------------------------------------------------------
+
+  @partial(jax.pmap, axis_name="batch", donate_argnums=(0, 1))
+  def update_fn(params, opt, rng, images, labels, step):
+    """Update step."""
+
+    measurements = {}
+
+    if config.get("mixup") and config.mixup.p:
+      rng, (images, labels), _ = u.mixup(rng, images, labels, **config.mixup)
+
+    # Get device-specific loss rng.
+    rng, rng_model = jax.random.split(rng, 2)
+    rng_model_local = jax.random.fold_in(rng_model, jax.lax.axis_index("batch"))
+
+    def loss_fn(params, images, labels):
+      logits, _ = model.apply(
+          {"params": flax.core.freeze(params)}, images,
+          train=True, rngs={"dropout": rng_model_local})
+      return getattr(u, config.get("loss", "sigmoid_xent"))(
+          logits=logits, labels=labels)
+
+    if config.get("GSAM", False):
```

Comment:
Since here we're specifically in the `gsam/train.py`, we can remove this config variable and `if` statement, and always execute the GSAM branch.